File: JsonPathParserLib/ParsedProperty.cs
namespace JsonPathParserLib
{
    public class ParsedProperty
    {
        public int StartPosition = -1;
        public int EndPosition = -1;
        public string Name = string.Empty;
        public string Value = string.Empty;
        public JsonPropertyType JsonPropertyType = JsonPropertyType.Unknown;
        public JsonValueType ValueType;
        public char PathDivider { get; private set; }

        public string Path
        {
            get => _path;
            set
            {
                _parentPath = null;
                _path = value;
            }
        }

        private string _path = string.Empty;

        public string ParentPath => _parentPath ?? (_parentPath = TrimPathEnd(Path, 1, PathDivider));
        private string _parentPath = null;

        public int RawLength
        {
            get
            {
                if (StartPosition == -1 || EndPosition == -1)
                    return -1;

                return EndPosition - StartPosition + 1;
            }
        }

        public int Depth
        {
            get
            {
                if (string.IsNullOrEmpty(_path))
                    return 0;

                return _path.Split(PathDivider).Length;
            }
        }

        public ParsedProperty(char pathDivider = '.')
        {
            PathDivider = pathDivider;
        }

        private static string TrimPathEnd(string originalPath, int levels, char pathDivider)
        {
            for (; levels > 0; levels--)
            {
                var pos = originalPath.LastIndexOf(pathDivider);
                if (pos > 0)
                    originalPath = originalPath.Substring(0, pos);
                else
                    return "";
            }

            return originalPath;
        }

        public override string ToString()
        {
            return Path;
        }
    }
}

File: typescript/ParsedProperty.ts
import { JsonPropertyType } from "./JsonPropertyType";
import { JsonValueType } from "./JsonValueTypes";

export default class ParsedProperty {

    public StartPosition: number = -1;
    public EndPosition: number = -1;
    public Name: string = "";
    public Value: string = "";
    public JsonPropertyType: JsonPropertyType = JsonPropertyType.Unknown;
    public ValueType: JsonValueType;
    public PathDivider: string = '.';

    public constructor(pathDivider: string) {
        this.PathDivider = pathDivider;
    }

    private _path: string = "";
    public set Path(value: string) {
        this._parentPath == null;
        this._path = value;
    }

    public get Path(): string {

        return this._path;
    }

    private _parentPath: string | null;
    public get ParentPath(): string {
        if ((this._parentPath == null)) {
            this._parentPath = ParsedProperty.TrimPathEnd(this._path, 1, this.PathDivider);
        }

        return this._parentPath;
    }

    public get RawLength(): number {
        if (((this.StartPosition == -1)
            || (this.EndPosition == -1))) {
            return -1;
        }

        return ((this.EndPosition - this.StartPosition)
            + 1);
    }

    private static TrimPathEnd(originalPath: string, levels: number, pathDivider: string): string {
        for (
            ; levels > 0; levels--) {
            let pos = originalPath.lastIndexOf(pathDivider);
            if ((pos >= 0)) {
                originalPath = originalPath.substring(0, pos);
            }
            else {
                break;
            }

        }

        return originalPath;
    }

    public ToString(): string {
        return this._path;
    }
}